clang/lib/DPCT/AnalysisInfo.cpp

```diff
@@ -399,6 +402,150 @@ void DpctFileInfo::postProcess() {
   }
 }
 
+class RnnBackwardFuncInfoBuilder {
+  std::vector<RnnBackwardFuncInfo> &RBFuncInfo;
+  std::vector<RnnBackwardFuncInfo> ValidBackwardDataFuncInfo;
+  std::vector<RnnBackwardFuncInfo> ValidBackwardWeightFuncInfo;
+  std::vector<std::shared_ptr<ExtReplacement>> Repls;
+  using InfoIter = std::vector<RnnBackwardFuncInfo>::iterator;
+
+public:
+  RnnBackwardFuncInfoBuilder(std::vector<RnnBackwardFuncInfo> &Infos)
+      : RBFuncInfo(Infos){};
+  bool isInputNotChanged(InfoIter Data, InfoIter Weight) {
+    for (auto RnnInput : Data->RnnInputDeclLoc) {
+      auto &RnnInputRefs =
+          DpctGlobalInfo::getRnnInputMap()[RnnInput][Data->FilePath];
+      for (auto &RnnInputRef : RnnInputRefs) {
+        if ((RnnInputRef > (Data->Offset + Data->Length - 1)) &&
+            RnnInputRef < Weight->Offset) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+  bool isInputSame(InfoIter Data, InfoIter Weight) {
+    for (unsigned InputIndex = 0; InputIndex < 3; InputIndex++) {
+      if (Data->RnnInputDeclLoc[InputIndex] !=
+          Weight->RnnInputDeclLoc[InputIndex]) {
+        return false;
+      }
+    }
+    return true;
+  }
+  bool isValidScopeAndOrder(InfoIter Data, InfoIter Weight) {
```

Comment:
please add comment to explain the function of isValidScopeAndOrder(),  give presudo code of dnn API pattern to be processed.